File: fibonacci/fibo.c

#include <stdio.h>
#include <string.h>


static int cache[1024];

static int
check_cache(int n)
{
  if (n == 0 || cache[n] > 0)
    return cache[n];
  return -1;
}

static int
nth_fibo(int n)
{
  int fib = check_cache(n);
  if (fib < 0) {
    fib = nth_fibo(n-1) + nth_fibo(n-2);
    cache[n] = fib;
  }
  return fib;
}

int
main(int argc, char const *argv[])
{
  FILE *file = fopen(argv[1], "r");
  int n;

  cache[1] = 1;

  while (fscanf(file, " %d ", &n) > 0)
    printf("%d\n", nth_fibo(n));
  return 0;
}

File: fibonacci/fibo.rb
module Fibonacci
  class Fibo
    def initialize
      @cache = Hash.new { |hash, n| hash[n] = caluculate_nth_fibo(n) }
      @cache[0] = 0
      @cache[1] = 1
    end

    def nth_fibo(n)
      @cache[n]
    end

    private
      def caluculate_nth_fibo(n)
        nth_fibo(n-1) + nth_fibo(n-2)
      end
  end

  def self.run(filename = ARGV[0])
    File.open(filename) do |file|
      fibo_calculator = Fibo.new
      file.each_line do |line|
        line.strip!
        next if line.empty?

        puts fibo_calculator.nth_fibo(line.to_i)
      end
    end
  end
end

Fibonacci.run if __FILE__ == $0